[Profiler] Add `tag` property to `Result`

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81319
* #80822
* #80797
* __->__ #81965
* #81895
* #81958

With a bit of template deduction we can make the variant tell us what type it is, and then we don't have to rely on (and maintain) a bunch of `isinstance` checks in Python.

Differential Revision: [D38066829](https://our.internmc.facebook.com/intern/diff/D38066829/)

Commit:
[Profiler] Add `tag` property to `Result`

With a bit of template deduction we can make the variant tell us what type it is, and then we don't have to rely on (and maintain) a bunch of `isinstance` checks in Python.

Differential Revision: [D38066829](https://our.internmc.facebook.com/intern/diff/D38066829/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Add `tag` property to `Result`"


With a bit of template deduction we can make the variant tell us what type it is, and then we don't have to rely on (and maintain) a bunch of `isinstance` checks in Python.

Differential Revision: [D38066829](https://our.internmc.facebook.com/intern/diff/D38066829/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Add `tag` property to `Result`"


With a bit of template deduction we can make the variant tell us what type it is, and then we don't have to rely on (and maintain) a bunch of `isinstance` checks in Python.

Differential Revision: [D38066829](https://our.internmc.facebook.com/intern/diff/D38066829/)

[ghstack-poisoned]

davidchencsl reviewed (approved):
LGTM! Neat and clean tricks as well.

Commit:
Update on "[Profiler] Add `tag` property to `Result`"


With a bit of template deduction we can make the variant tell us what type it is, and then we don't have to rely on (and maintain) a bunch of `isinstance` checks in Python.

Differential Revision: [D38066829](https://our.internmc.facebook.com/intern/diff/D38066829/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Add `tag` property to `Result`"


With a bit of template deduction we can make the variant tell us what type it is, and then we don't have to rely on (and maintain) a bunch of `isinstance` checks in Python.

Differential Revision: [D38066829](https://our.internmc.facebook.com/intern/diff/D38066829/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Add `tag` property to `Result`"


With a bit of template deduction we can make the variant tell us what type it is, and then we don't have to rely on (and maintain) a bunch of `isinstance` checks in Python.

Differential Revision: [D38066829](https://our.internmc.facebook.com/intern/diff/D38066829/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Add `tag` property to `Result`"


With a bit of template deduction we can make the variant tell us what type it is, and then we don't have to rely on (and maintain) a bunch of `isinstance` checks in Python.

Differential Revision: [D38066829](https://our.internmc.facebook.com/intern/diff/D38066829/)

[ghstack-poisoned]

robieta commented:
@pytorchbot merge -l

robieta commented:
@pytorchbot merge -l

robieta commented:
@pytorchbot merge
